agent/structs/config_entry_gateways.go

```diff
@@ -679,3 +679,127 @@ func (g *GatewayService) Clone() *GatewayService {
 		ServiceKind:  g.ServiceKind,
 	}
 }
+
+// APIGatewayConfigEntry manages the configuration for an API gateway
+// with the given name.
+type APIGatewayConfigEntry struct {
+	Kind string
+	Name string
+
+	Meta               map[string]string `json:",omitempty"`
+	acl.EnterpriseMeta `hcl:",squash" mapstructure:",squash"`
+	RaftIndex
+}
+
+func (e *APIGatewayConfigEntry) GetKind() string {
+	return APIGateway
+}
+
+func (e *APIGatewayConfigEntry) GetName() string {
+	if e == nil {
+		return ""
+	}
+	return e.Name
+}
+
+func (e *APIGatewayConfigEntry) GetMeta() map[string]string {
+	if e == nil {
+		return nil
+	}
+	return e.Meta
+}
+
+func (e *APIGatewayConfigEntry) Normalize() error {
+	return nil
+}
+
+func (e *APIGatewayConfigEntry) Validate() error {
+	return nil
+}
+
+func (e *APIGatewayConfigEntry) CanRead(authz acl.Authorizer) error {
+	var authzContext acl.AuthorizerContext
+	e.FillAuthzContext(&authzContext)
+	return authz.ToAllowAuthorizer().ServiceReadAllowed(e.Name, &authzContext)
+}
+
+func (e *APIGatewayConfigEntry) CanWrite(authz acl.Authorizer) error {
+	var authzContext acl.AuthorizerContext
+	e.FillAuthzContext(&authzContext)
+	return authz.ToAllowAuthorizer().MeshWriteAllowed(&authzContext)
+}
+
+func (e *APIGatewayConfigEntry) GetRaftIndex() *RaftIndex {
+	if e == nil {
+		return &RaftIndex{}
+	}
+	return &e.RaftIndex
+}
+
+func (e *APIGatewayConfigEntry) GetEnterpriseMeta() *acl.EnterpriseMeta {
+	if e == nil {
+		return nil
+	}
+	return &e.EnterpriseMeta
+}
+
+// BoundAPIGatewayConfigEntry manages the configuration for a bound API
+// gateway with the given name.
+type BoundAPIGatewayConfigEntry struct {
+	Kind string
+	Name string
+
+	Meta               map[string]string `json:",omitempty"`
+	acl.EnterpriseMeta `hcl:",squash" mapstructure:",squash"`
+	RaftIndex
+}
+
+func (e *BoundAPIGatewayConfigEntry) GetKind() string {
+	return APIGateway
+}
+
+func (e *BoundAPIGatewayConfigEntry) GetName() string {
+	if e == nil {
+		return ""
+	}
+	return e.Name
+}
+
+func (e *BoundAPIGatewayConfigEntry) GetMeta() map[string]string {
+	if e == nil {
+		return nil
+	}
+	return e.Meta
+}
+
+func (e *BoundAPIGatewayConfigEntry) Normalize() error {
+	return nil
+}
+
+func (e *BoundAPIGatewayConfigEntry) Validate() error {
+	return nil
+}
+
+func (e *BoundAPIGatewayConfigEntry) CanRead(authz acl.Authorizer) error {
+	var authzContext acl.AuthorizerContext
+	e.FillAuthzContext(&authzContext)
+	return authz.ToAllowAuthorizer().ServiceReadAllowed(e.Name, &authzContext)
+}
+
+func (e *BoundAPIGatewayConfigEntry) CanWrite(authz acl.Authorizer) error {
+	return acl.PermissionDeniedError{Cause: "only writeable by controller"}
```

Comment:
Correct. The test framework I plugged our ACL stuff into specifically expects a `acl.PermissionDeniedError` vs the vanilla `error` interface that I was using previously.

Reply:
FWIW, it's pretty common to return a `PermissionDeniedError` with only the `Cause` set

<img width="1060" alt="image" src="https://user-images.githubusercontent.com/3476400/207444484-03839875-4133-4c97-9fbd-2c129e04cdcd.png">

Reply:
After diving deeper here, I think only populating the `Cause` is the right move. Since various API assertions expect a `PermissionDeniedError`, we have to return that type; however, as soon as you assign a `Cause` to the error - which we have to do in order to avoid the default error message implying your ACLs are just incorrect - the rest of the information in the error is no longer visible. This is hinted at in the docstrings for `PermissionDeniedError.Error()` [here](https://github.com/hashicorp/consul/blob/bde57c0dd0670a4a9430e29c08ff03a52e614886/acl/errors.go#L79-L111). Seems like there's a desire to reconcile the different use cases for `PermissionDeniedError` in the future, but I don't believe this PR is the right place to do that.